src/watch_state.rs

```diff
@@ -42,33 +42,50 @@ impl WatchState {
         // https://docs.rs/watchexec/latest/watchexec/action/enum.Outcome.html
         let outcome = match (effect, *state) {
             (Effect::Exit, _) => {
-                Outcome::if_running(Outcome::both(Outcome::Stop, Outcome::Exit), Outcome::Exit)
+                Outcome::if_running(Outcome::both(stop_outcome(), Outcome::Exit), Outcome::Exit)
             }
             (Effect::ChildProcessFailed, _) => {
-                Outcome::if_running(Outcome::Stop, Outcome::DoNothing)
+                Outcome::if_running(stop_outcome(), Outcome::DoNothing)
             }
             (Effect::ChildProcessCompleted, State::Building) => {
                 *state = State::Running;
                 self.restart_outcome(PREVENT_CLEAR)
             }
-            (Effect::ChildProcessCompleted, State::Running) => Outcome::DoNothing,
+            (Effect::ChildProcessCompleted, State::Running) => {
+                *state = State::Building;
+                self.restart_outcome(PREVENT_CLEAR)
+            }
+            (Effect::ChildProcessCompleted, State::WaitingForSpinUpToExit) => {
+                *state = State::Building;
+                self.restart_outcome(PREVENT_CLEAR)
+            }
             (Effect::ManifestChange, State::Building) => self.restart_outcome(ALLOW_CLEAR),
+            (Effect::ManifestChange, State::WaitingForSpinUpToExit) => Outcome::DoNothing,
             (Effect::ManifestChange, State::Running) => {
                 if !self.skip_build {
-                    *state = State::Building;
+                    *state = State::WaitingForSpinUpToExit;
+                    Outcome::if_running(stop_outcome(), Outcome::DoNothing)
+                } else {
+                    self.restart_outcome(ALLOW_CLEAR)
                 }
-                self.restart_outcome(ALLOW_CLEAR)
             }
             (Effect::SourceChange, State::Building) => self.restart_outcome(ALLOW_CLEAR),
```

Comment:
Does this not cause a race condition. Since we now use SIGINT the build will take awhile to finish and so a `Effect::ChildProcessCompleted` will come while the build is running and that will prematurely kill the build and jump to the running step?

I've been staring at this a long time and struggling to think it through so I could very well be wrong.

Reply:
Yeah I am going to write out all the start scenarios "inline" and think through whether they are correct.